src/lib/shell/commands/Dns.cpp

```diff
@@ -0,0 +1,141 @@
+                        ChipLogValueX64(nodeData.mPeerId.GetFabricId()), ChipLogValueX64(nodeData.mPeerId.GetNodeId()));
+        streamer_printf(streamer_get(), "   IP address: %s\n", nodeData.mAddress.ToString(ipAddressBuf));
+        streamer_printf(streamer_get(), "   Port: %d\n", nodeData.mPort);
+    }
+
+    void OnNodeIdResolutionFailed(const PeerId & peerId, CHIP_ERROR error) override {}
+
+    void OnNodeDiscoveryComplete(const chip::Mdns::DiscoveredNodeData & nodeData) override
+    {
+        if (!nodeData.IsValid())
+        {
+            streamer_printf(streamer_get(), "DNS browse failed - not found valid services \n");
+            return;
+        }
+        streamer_printf(streamer_get(), "DNS browse succeeded: \n");
+        streamer_printf(streamer_get(), "   Hostname: %s\n", nodeData.hostName);
+        streamer_printf(streamer_get(), "   Vendor ID: %d\n", nodeData.vendorId);
+        streamer_printf(streamer_get(), "   Product ID: %d\n", nodeData.productId);
+        streamer_printf(streamer_get(), "   Long discriminator: %d\n", nodeData.longDiscriminator);
+        streamer_printf(streamer_get(), "   Device type: %d\n", nodeData.deviceType);
+        streamer_printf(streamer_get(), "   Device name: %s\n", nodeData.deviceName);
+        streamer_printf(streamer_get(), "   Commissioning mode: %d\n", nodeData.commissioningMode);
+        streamer_printf(streamer_get(), "   IP addresses:\n");
+        for (uint8_t i = 0; i < nodeData.kMaxIPAddresses; i++)
+        {
+            if (nodeData.ipAddress[i] != chip::Inet::IPAddress::Any)
+                streamer_printf(streamer_get(), "      %s\n", nodeData.ipAddress[i].ToString(ipAddressBuf));
+        }
+    }
+
+private:
+    char ipAddressBuf[chip::Inet::kMaxIPAddressStringLength];
+};
+
+static DnsShellResolverDelegate sDnsShellResolverDelegate;
+
+CHIP_ERROR DnsHelpHandler(int argc, char ** argv)
+{
+    sShellDnsSubcommands.ForEachCommand(PrintCommandHelp, nullptr);
+    return CHIP_NO_ERROR;
+}
+
+static CHIP_ERROR ResolveHandler(int argc, char ** argv)
+{
+    streamer_printf(streamer_get(), "Resolving ...\r\n");
+
+    VerifyOrReturnError(argc == 2, CHIP_ERROR_INVALID_ARGUMENT);
+
+    PeerId peerId;
+    peerId.SetFabricId(strtoull(argv[0], NULL, 10));
+    peerId.SetNodeId(strtoull(argv[1], NULL, 10));
+
+    return chip::Mdns::Resolver::Instance().ResolveNodeId(peerId, Inet::kIPAddressType_Any);
+}
+
+static CHIP_ERROR BrowseHandler(int argc, char ** argv)
+{
+    streamer_printf(streamer_get(), "Browsing ...\r\n");
+
+    return chip::Mdns::Resolver::Instance().FindCommissionableNodes();
+}
+
+static CHIP_ERROR DnsHandler(int argc, char ** argv)
+{
+    if (!isResolverStarted)
+    {
+        chip::Mdns::Resolver::Instance().StartResolver(&chip::DeviceLayer::InetLayer, chip::Mdns::kMdnsPort);
+        chip::Mdns::Resolver::Instance().SetResolverDelegate(&sDnsShellResolverDelegate);
```

Comment:
I am generally concerned at all of the code we're building up that enforces the singleton pattern for the mDNS module.  I'm not sure why this is necessary.  But it's going to force platforms that cannot tolerate this to interact with mDNS outside of the sdk, reducing the usefulness of the sdk's mDNS code in such cases.

And I think such scenarios are likely.  All it would take for this to break is a single case where multiple separate, but concurrent discovery operations need to occur at the same time.  I can think of several cases where a node would want to do that.

Reply:
@msandstedt Can you provide any use cases where using the singleton pattern for mDNS can be limiting? I always thought that the biggest issue in our codebase is that we allow a single delegate in certain components, so it's hard to spawn multiple jobs of the same kind in parallel. I thought that some pattern like:
```
struct ResolveRequest
{
  NodeId mNodeId;
  // ... other request params
  OnComplete mOnCompleteCallback;
  ResolveRequest* mNext;

  void Complete(ResolveResult* result);
  void Cancel();
};
```
might be a good idea to solve this issue. But I would welcome any arguments for multiple instances of the mDNS server in CHIP as well.